events/UserAllocatedEvent.h

```diff
@@ -0,0 +1,435 @@
+/*
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef USER_ALLOCATED_EVENT_H
+#define USER_ALLOCATED_EVENT_H
+
+#include "events/EventQueue.h"
+#include "platform/mbed_assert.h"
+#include "platform/mbed_atomic.h"
+
+namespace events {
+/**
+ * \addtogroup events-public-api Events
+ * \ingroup mbed-os-public
+ * @{
+ */
+template <typename F, typename A>
+class UserAllocatedEvent;
+
+/**
+ * \defgroup events_Event UserAllocatedEvent class
+ * @{
+ */
+
+/** UserAllocatedEvent
+ *
+ *  Representation of an static event for fine-grain dispatch control.
+ *
+ *  UserAllocatedEvent provides mechanism for event posting and dispatching
+ *  without utilization of queue internal memory. It embeds all underlying
+ *  event data and doesn't require any memory allocation while posting and dispatching.
+ *  All of these makes it cannot fail due to memory exhaustion while posting
+ *
+ * Usage:
+ * @code
+ *  #include "mbed.h"
+ *
+ *  void handler(int data) { ... }
+ *
+ *  class Device {
+ *     public:
+ *     void handler(int data) { ... }
+ *  };
+ *
+ *  Device dev;
+ *  // Create event not bounded to any queue
+ *  static auto e1 = make_user_allocated_event(&dev, Device::handler, 2);
+ *  static auto e2 = make_user_allocated_event(handler, 3);
+ *
+ *  int main()
+ *  {
+ *    // queue with not internal storage for dynamic events
+ *    // accepts only user allocated events
+ *    EventQueue queue(1);
```

Comment:
That still requires malloc (even if of 1). Could we have a special mode that doesn't pull the malloc in at all?

Reply:
Yes, you should be able to just do queue(0) if intending only to use static events.

(With current code, you could do `queue(1, &my_single_char)` to avoid the malloc.)

Reply:
or do it this way
```diff 
 EventQueue::EventQueue(unsigned event_size, unsigned char *event_pointer)
 {
+    if (event_size == 0) {
+        static uint8_t dummy;
+        equeue_create_inplace(&_equeue, event_size, &dummy);
+    }
     if (!event_pointer) {
         equeue_create(&_equeue, event_size);
     } else {
```

Reply:
Looks good, assuming it never touches `dummy`. And if it REALLY doesn't ever touch dummy, you could, in principle just pass `&_equeue` as your buffer instead, saving 1 byte of RAM (in all images - that static dummy gets reserved whatever).

Reply:
I encountered that doing it via `static uint8_t dummy;` cause some strange run time errors.
e.g. this assert  https://github.com/ARMmbed/mbed-os/blob/98c0fd06be9aa0a97125bcd947ebbd5cd3df6e94/platform/SingletonPtr.h#L139
So for now dummy is regular EventQueue member.
I think that 1B more  per queue is not much. 
How many queues utilize typical program?

Reply:
That sounds dangerous. As that assert notes, if it's triggered, someone has corrupted memory. Which suggests that your attempt to do that has made it run off the end of `_dummy`. It may still be doing that, with the dummy elsewhere, and you're not noticing.

Is there some assumption that the buffer is at least some size, and it's writing in a fixed header?

You could catch that with a watchpoint on `_dummy + 1`.

Reply:
fixed. The problem was in queue buffer alignment
```c++
int equeue_create_inplace(equeue_t *q, size_t size, void *buffer)
{
    // setup queue around provided buffer
    // ensure buffer and size are aligned
    if (size >= sizeof(void *)) {
        q->buffer = (void *)(((uintptr_t) buffer + sizeof(void *) -1) & ~(sizeof(void *) -1));
        size -= (char *) q->buffer - (char *) buffer;
        size &= ~(sizeof(void *) -1);
    } else {
        // don't align when size less then pointer size
        // e.g. static queue (size == 1)
        q->buffer = buffer;
    }
```

Reply:
Cool, but why is it accessing the buffer at all, even after alignment?